docs/framework/react/guide/type-safety.md

```diff
@@ -115,6 +115,47 @@ const router = createRouter({
 
 As your application scales, TypeScript check times will naturally increase. There are a few things to keep in mind when your application scales to keep your TS check times down.
 
+### Only infer types you need
+
+A great pattern with client side data caches (TanStack Query etc.) is to prefetch data. For example with TanStack Query you might have a route which calls `ensureQueryData` in a `loader`.
+
+```tsx
+export const Route = createFileRoute('/posts/$postId/deep')({
+  loader: ({ context: { queryClient }, params: { postId } }) =>
+    queryClient.ensureQueryData(postQueryOptions(postId)),
+  errorComponent: PostErrorComponent as any,
+  component: PostDeepComponent,
+})
+
+function PostDeepComponent() {
+  const params = Route.useParams()
+  const data = useSuspenseQuery(postQueryOptions(params.postId))
+
+  return <></>
+}
+```
+
+This may look fine and for small route trees you may not notice any TS performance issues. However in this case TS has to infer the loader return type despite it never beng used in your route. If the loader data is a complex type and many routes prefetch in this manner it can slow down editor performance. In this case the change is simple and you can let `Promise<void>` be inferred
```

Comment:
- ... infer the loader's return type, despite ...
- ... data is a complex type with many routes ...
- In this case, the change is quite simple and let's type of `Promise<void>` be inferred.